kaitaistruct.py

```diff
@@ -15,51 +20,171 @@
 __version__ = '0.9'
 
 
-class KaitaiStruct(object):
-    def __init__(self, stream):
-        self._io = stream
+class _KaitaiStruct_:
+    __slots__ = ("_io", "_parent", "_root")
+    def __init__(self, _io:"KaitaiStream", _parent:typing.Optiohal["_KaitaiStruct"]=None, _root:typing.Optiohal["_KaitaiStruct"]=None):
+        self._io = _io
+        self._parent = _parent
+        self._root = _root if _root else self
+
+
+class KaitaiParser(ABC):
+    __slots__ = () # in fact this class not meant to be instantiated
+
+    @abstractmethod
+    @classmethod
+    def parse(cls, struct:_KaitaiStruct_, io:"KaitaiStream"):
+        raise NotlmplementedError()
+
+
+class _KaitaiStruct(_KaitaiStruct_):
+    __slots__ = ()
+    _parser = None  # :KaitaiParser
+
+    def _read(self):
+        self.__class__._parser.parse(self, self._io)
+
+
+class KaitaiStruct(_KaitaiStruct):
+    __slots__ = ("_shouldExit",)
+    def __init__(self, io: "KaitaiStream"):
+        super.__init__(io)
+        self._shouldExit = False
 
     def __enter__(self):
+        self._shouldExit = not stream.is_entered
+        if self._shouldExit:
+            self._io.__enter__()
```

Comment:
I'm wondering if we should assume that `__enter__` on IO streams never does anything - it would simplify our own `__enter__` implementations a lot. I think this is a safe assumption, because you can always use streams with manual `close` calls instead of `with`.

Reply:
IMHO - `close` should be removed and constructing shouldn't mean `__enter__`ing in python 4.

Reply:
> IMHO - `close` should be removed in favour of `__exit__` and constructing shouldn't mean `__enter__`ing in python 4 for some built-in types.

That's unlikely to happen. The general convention in the stdlib seems to be that you're never *forced* to use `with` (or `__enter__`/`__exit__`), because there are always other ways to do what `with` would do (such as `close` on streams).

> Yes, I am mad enough to assumme that there will be python 4 fixing the mistakes of python 3 somewhen. And I sincerely hope it will happen. In fact I have a wishlist of the changes it is impossible to implement without breaking compatibility I wanna see in python.

That will never happen. I can't find a quote for this right now, but the core devs have said that Python 4 won't be a big release that breaks everything like Python 3. I don't think that the devs (or the community) are interested in another messy incompatible version switch like Python 2 to 3 was.